### PR TITLE
ignite-2190

### PR DESCRIPTION
**Pull request by nizhikov**
* fix issue
* Added GridCacheEmptyScanQueryTest to reproduce bug described in task
* Minor fixes in IgniteProcessProxy and GridAbstractTest for a possibility to modify separate jvm grid classpath
* Added ignite-2190-1.0.jar to binary dependencies(source can be obtained from https://github.com/nizhikov/ignite-2190-binary)

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryObjectImpl.java [hunk 1]
```diff
@@ -789,8 +789,14 @@ else if (fieldOffsetLen == BinaryUtils.OFFSET_2)
      * @return Object.
      */
     private Object deserializeValue(@Nullable CacheObjectValueContext coCtx) {
-        BinaryReaderExImpl reader = reader(null, coCtx != null ?
-            coCtx.kernalContext().config().getClassLoader() : ctx.configuration().getClassLoader(), true);
+        ClassLoader clsLdr;
+        if (coCtx == null)
```
Comment (daradurvs) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryObjectImpl.java]: New line requered

File: modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryObjectImpl.java [hunk 2]
```diff
@@ -789,8 +789,14 @@ else if (fieldOffsetLen == BinaryUtils.OFFSET_2)
...
+            clsLdr = ctx.configuration().getClassLoader();
+        else if (coCtx.kernalContext().config().isPeerClassLoadingEnabled())
+            clsLdr = coCtx.kernalContext().cache().context().deploy().globalLoader();
+        else
+            clsLdr = coCtx.kernalContext().config().getClassLoader();
+        BinaryReaderExImpl reader = reader(null, clsLdr, true);
```
Comment (daradurvs) [inline modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryObjectImpl.java]: New line

File: modules/core/src/test/binaries/repo/org/apache/ignite/binary/ignite-2190/1.0/ignite-2190-1.0.pom [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (daradurvs) [inline modules/core/src/test/binaries/repo/org/apache/ignite/binary/ignite-2190/1.0/ignite-2190-1.0.pom]: Apache license header requered

File: modules/core/src/test/binaries/repo/org/apache/ignite/binary/ignite-2190/maven-metadata-local.xml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<metadata>
```
Comment (daradurvs) [inline modules/core/src/test/binaries/repo/org/apache/ignite/binary/ignite-2190/maven-metadata-local.xml]: Apache license header requered

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
+package org.apache.ignite.internal.processors.cache.query;
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java:18]: Apache license header requered

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+import org.apache.ignite.lang.IgniteBiPredicate;
+import org.apache.ignite.test.ignite2190.Employee;
+import org.apache.ignite.test.ignite2190.EmployeePredicate;
+import org.apache.ignite.test.ignite2190.ObjectPredicate;
+import org.apache.ignite.testframework.config.GridTestProperties;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.testframework.junits.multijvm.IgniteProcessProxy;
+
+/**
+ * @author NIzhikov
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: @author isn't allowed, just the class description is needed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+ */
+public class GridCacheEmptyScanQueryTest extends GridCommonAbstractTest {
+    /**
+     * jar name contains Employee class
+     */
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: Oneline javadoc style requered and 'dot' in the end

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    private static final String IGNITE_2190_1_0_JAR = "ignite-2190-1.0.jar";
+
+    @Override protected boolean isMultiJvm() {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: /** {@inheritDoc} */

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String gridName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(gridName);
+        cfg.setClientMode(false);
+        cfg.setPeerClassLoadingEnabled(true);
+        return cfg;
+    }
+
+    public void testEmptyScanQueryWithOptimizedMarshaller() throws Exception {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc is necessary. For example:
/**
     * @throws Exception If failed.
     */

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        runEmptyScanQuery(new ScanQuery<>(), 2,
+            "org.apache.ignite.internal.marshaller.optimized.OptimizedMarshaller");
+    }
+
+    public void testEmptyScanQueryWithBinaryMarshaller() throws Exception {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        runEmptyScanQuery(new ScanQuery<>(), 2,
+            "org.apache.ignite.internal.binary.BinaryMarshaller");
+    }
+
+    public void testEmptyScanQuery2WithOptimizedMarshaller() throws Exception {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        runEmptyScanQuery(new ScanQuery<Integer, Employee>(new ObjectPredicate()), 2,
+            "org.apache.ignite.internal.marshaller.optimized.OptimizedMarshaller");
+    }
+
+    public void testEmptyScanQuery2WithBinaryMarshaller() throws Exception {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        runEmptyScanQuery(new ScanQuery<Integer, Employee>(new ObjectPredicate()), 2,
+            "org.apache.ignite.internal.binary.BinaryMarshaller");
+    }
+
+    public void testEmptyScanQuery3WithOptimizedMarshaller() throws Exception {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        runEmptyScanQuery(new ScanQuery<>(new EmployeePredicate()), 2,
+            "org.apache.ignite.internal.marshaller.optimized.OptimizedMarshaller");
+    }
+
+    public void testEmptyScanQuery3WithBinaryMarshaller() throws Exception {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+        runEmptyScanQuery(new ScanQuery<>(new EmployeePredicate()), 2,
+            "org.apache.ignite.internal.binary.BinaryMarshaller");
+    }
+
+    public void runEmptyScanQuery(ScanQuery<Integer, Employee> query, int expectedSize,
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc.
query -> qry
expectedSize -> expSz
https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+            assertEquals("Size of result of always true ScanQuery should be 2", expectedSize,
+                cache.query(query).getAll().size());
+        }
+        finally {
+            stopAllGrids();
+        }
+    }
+
+    private IgniteConfiguration getLocalConfiguration() throws Exception {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+        final IgniteConfiguration cfg = getConfiguration();
+        cfg.setClientMode(true);
+        return cfg;
+    }
+
+    @Override protected IgniteProcessProxy igniteProcessProxy(IgniteConfiguration cfg, Ignite locNode,
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: javadoc

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+        return new IgniteProcessProxy(cfg, log, grid(0)) {
+            @Override protected Collection<String> filteredJvmArgs(IgniteConfiguration cfg) {
+                Collection<String> defaultFilteredJvmArgs = super.filteredJvmArgs(cfg);
+                List<String> filteredJvmArgs = new ArrayList<>();
+
+                boolean classpathFound = false;
+                Iterator<String> iDefaultFilteredJvmArgs = defaultFilteredJvmArgs.iterator();
+                while (iDefaultFilteredJvmArgs.hasNext()) {
+                    String arg = iDefaultFilteredJvmArgs.next();
+                    if ("-cp".equals(arg) || "-classpath".equals(arg)) {
```
Comment (daradurvs) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: new line

**Review by daradurvs — CHANGES_REQUESTED** [review]
Need to make code style according to https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+    @Override protected boolean isMultiJvm() {
+        return true;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String gridName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(gridName);
+        cfg.setClientMode(false);
+        cfg.setPeerClassLoadingEnabled(true);
+        return cfg;
```
Comment (BiryukovVA) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: new line

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+        finally {
+            stopAllGrids();
+        }
+    }
+
+    /** Return configuration for grid running in same jvm. */
+    private IgniteConfiguration getLocalConfiguration() throws Exception {
+        final IgniteConfiguration cfg = getConfiguration();
+        cfg.setClientMode(true);
+        return cfg;
```
Comment (BiryukovVA) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: new line

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+
+                return filteredJvmArgs;
+            }
+
+            /** Excluding ignite-2190-1.0.jar so Employee class become invisible for a remote node. */
+            String excludeIgnite2190JarFromClasspath(String classpath) {
+                final String pathSeparator = System.getProperty("path.separator");
+                final String[] classpathArr = classpath.split(pathSeparator);
+                classpath = "";
+                for (String aClasspathArr : classpathArr) {
```
Comment (BiryukovVA) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: new line

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+                    if (!aClasspathArr.contains(IGNITE_2190_1_0_JAR)) {
+                        classpath += (classpath.length() > 0 ? pathSeparator : "") + aClasspathArr;
```
Comment (BiryukovVA) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: one line "if" without "{}"

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+                    }
+                }
+                return classpath;
```
Comment (BiryukovVA) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/GridCacheEmptyScanQueryTest.java]: new line

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteProcessProxy.java [hunk 24]
```diff
@@ -192,6 +176,31 @@ public IgniteProcessProxy(IgniteConfiguration cfg, IgniteLogger log, Ignite locJ
...
+        else
+            filteredJvmArgs.add("-D" + IgniteTestResources.MARSH_CLASS_NAME + "=" + BinaryMarshaller.class.getName());
+
+        for (String arg : U.jvmArgs()) {
+            if (arg.startsWith("-Xmx") || arg.startsWith("-Xms") ||
+                arg.startsWith("-cp") || arg.startsWith("-classpath") ||
+                (marsh != null && arg.startsWith("-D" + IgniteTestResources.MARSH_CLASS_NAME)))
+                filteredJvmArgs.add(arg);
+        }
+        return filteredJvmArgs;
```
Comment (BiryukovVA) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteProcessProxy.java]: new line

**Review by BiryukovVA — CHANGES_REQUESTED** [review]
Need to make code style according to https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteProcessProxy.java (hunk 24 above)
Comment (BiryukovVA) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/multijvm/IgniteProcessProxy.java]: rerurn need to start from a new line.

**Review by BiryukovVA — CHANGES_REQUESTED** [review]

**Pull request closed without merging**